### PR TITLE
Algolia search

### PR DESCRIPTION
**Pull request by Tehsurfer**
Switch to using Algolia search for facets and searching. 

We still require a different index for scaffold and plot information, this is done via the endpoint here:
https://github.com/nih-sparc/sparc-api/pull/59

A live demo of this can be viewed at:
https://algolia-maps.herokuapp.com/maps

### CODE REVIEW

**Comment by alan-wu** [comment]
@Tehsurfer It looks good in general but I found some bugs, please take a look. I will review the codes once these bugs are resolved.

1. The filter is reset after clicking on the flatmap's marker then click to different page of result on the sidebar.
2. Choose scaffold in Experimental approach -> 16 results, go to second page -> no result found.
3. Show all should be selected by default and when nothing else is selected.
4. When show all on any filter is selected, no result is found.

File: src/algolia/utils.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  'organisms.primary.species.name' : 'Species',
+  'item.modalities.keyword' : 'Experimental Approach',
+  'attributes.subject.sex.value' : 'Sex',
+  'attributes.subject.ageCategory.value' : 'Age Categories',
+}
+
+export const getAlgoliaFacets = function(algoliaIndex, propPathMapping) {
+  const map = new Map(Object.entries(propPathMapping));
+  const facetPropPaths = Array.from(map.keys() );
+  var facetData = []
```
Comment (alan-wu) [inline src/algolia/utils.js:14]: I found multiple occurrences of vars and lets. It is better and safer to use let to avoid problem with scopes.

Comment (Tehsurfer) [reply src/algolia/utils.js:14]: Ah yeah good catch! will fix

**Comment by Tehsurfer** [comment]
Thanks so much for the review @alan-wu! 

Good finds will fix these

File: src/algolia/algolia.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import algoliasearch from 'algoliasearch'
+
+// export `createAlgoliaClient` to use it in page components
+export default function createAlgoliaClient() {
+  const client = algoliasearch(
+    process.env.VUE_APP_ALGOLIA_ID,
```
Comment (alan-wu) [inline src/algolia/algolia.js:6]: These should probably be passed into it using prop.

File: src/algolia/utils.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  var facetId = 0
+  return algoliaIndex
+    .search('', {
+      sortFacetValuesBy: 'alpha',
+      facets: facetPropPaths
+    })
+    .then(response => {
+      facetPropPaths.map((facetPropPath) => {
```
Comment (alan-wu) [inline src/algolia/utils.js:22]: It is better to use forEach as the returned array from map is not use.

File: src/algolia/utils.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        var children = []
+        const responseFacets = response.facets
+        if (responseFacets === undefined) {return}
+        const responseFacetChildren =
+          responseFacets[facetPropPath] == undefined
+            ? {}
+            : responseFacets[facetPropPath]
+        Object.keys(responseFacetChildren).map(facet => {
```
Comment (alan-wu) [inline src/algolia/utils.js:30]: Same as above.

File: src/components/SearchFilters.vue [hunk 5]
```diff
@@ -103,109 +112,180 @@ export default {
...
+     * i.e. (color:blue OR color:red) AND (shape:circle OR shape:red) */
+    getFilters(selectedFacetArray) {
+      if (selectedFacetArray === undefined) {
+        return undefined;
+      }
+      var filters = "NOT item.published.status:embargo";
+
+      filters = `(${filters}) AND `;
+      const facetPropPaths = Object.keys(facetPropPathMapping);
+      window.propPathsStart = facetPropPaths;
```
Comment (alan-wu) [inline src/components/SearchFilters.vue:220]: Remove this line please.

File: src/components/SearchFilters.vue [hunk 6]
```diff
@@ -103,109 +112,180 @@ export default {
...
+      return new Promise(resolve => {
+        this.algoliaIndex
+        .search('', {facets:['*'],filters:filter})
+        .then(response => {
+          let searchData = {
+            items: response.hits,
+            total: response.nbHits
+          }
+          console.log(searchData)
+          window.searchData = searchData
```
Comment (alan-wu) [inline src/components/SearchFilters.vue:251]: Remove this line please.

File: src/components/SearchFilters.vue [hunk 7]
```diff
@@ -103,109 +112,180 @@ export default {
...
+    algoliaSearch: function(filter) {
+      return new Promise(resolve => {
+        this.algoliaIndex
+        .search('', {facets:['*'],filters:filter})
+        .then(response => {
+          let searchData = {
+            items: response.hits,
+            total: response.nbHits
+          }
+          console.log(searchData)
```
Comment (alan-wu) [inline src/components/SearchFilters.vue:250]: Remove this line please.

File: src/components/SearchFilters.vue [hunk 8]
```diff
@@ -327,19 +394,18 @@ export default {
...
-    }
-  },
-  created: function() {
-    //Create non-reactive local variables
-    this.facetEndpoint = "get-facets/";
+    },
   },
-  mounted: function() {
+  mounted: function () {
+    window.envVars = this.envVars
```
Comment (alan-wu) [inline src/components/SearchFilters.vue:400]: Remove this one please.

**Comment by alan-wu** [comment]
@Tehsurfer I have a better look at the logic in the codes, looks like clicking on the paigination will run the filter search using the sparc-api instead of Algolia search. Setting the filter programmatically - using the "openSearch" method also triggers the filter search on the API instead of Algolia.
The number of results returned from the call to multiple dois is capped at 20, this has something to do with the default number of return in the request body.

**Pull request merged**